modules/flyte/roles.tf

```diff
@@ -1,30 +1,56 @@
-      "Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey/action"
+    data_actions = [
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read",
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write",
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/add/action",
     ]
+  }
+}
+
+resource "azurerm_role_definition" "flyte_data_role" {
+  name  = "${var.deploy_id}-flyte-data"
+  scope = azurerm_storage_container.flyte_data.resource_manager_id
+  permissions {
     data_actions = [
       "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read",
       "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write",
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/add/action",
     ]
   }
-  assignable_scopes = [
-    azurerm_storage_container.flyte_metadata.resource_manager_id,
-    azurerm_storage_container.flyte_data.resource_manager_id,
-  ]
 }
 
-resource "azurerm_role_assignment" "flyte_metadata_access" {
+# Because the Get User Delegation Key operation acts at the level of the storage account, the
+# Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey action must be scoped at the level
+# of the storage account, the resource group, or the subscription.
+# https://learn.microsoft.com/en-us/rest/api/storageservices/get-user-delegation-key
+#
+resource "azurerm_role_definition" "flyte_sas_role" {
+  name  = "${var.deploy_id}-flyte-sas"
+  scope = azurerm_storage_account.flyte.id
+  permissions {
+    actions = [
+      "Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey/action"
+    ]
+  }
+}
+
+resource "azurerm_role_assignment" "flyte_metadata_role_assignment" {
   scope              = azurerm_storage_container.flyte_metadata.resource_manager_id
-  role_definition_id = azurerm_role_definition.flyte_storage_access.role_definition_resource_id
-  principal_id       = azuread_group.flyte_metadata_access.object_id
+  role_definition_id = azurerm_role_definition.flyte_metadata_role.role_definition_resource_id
+  principal_id       = azuread_group.flyte_metadata_group.object_id
 }
 
-resource "azurerm_role_assignment" "flyte_data_access" {
+resource "azurerm_role_assignment" "flyte_data_role_assignment" {
   scope              = azurerm_storage_container.flyte_data.resource_manager_id
-  role_definition_id = azurerm_role_definition.flyte_storage_access.role_definition_resource_id
-  principal_id       = azuread_group.flyte_data_access.object_id
+  role_definition_id = azurerm_role_definition.flyte_data_role.role_definition_resource_id
+  principal_id       = azuread_group.flyte_data_group.object_id
+}
+
+resource "azurerm_role_assignment" "flyte_sas_role_assignment" {
+  scope              = azurerm_storage_account.flyte.id
+  role_definition_id = azurerm_role_definition.flyte_sas_role.role_definition_resource_id
```

Comment:
Azure's mapping of group -> role -> permissions is a bit confusing to me. How does the `flyte_sas_group` that contains the `flyte_dataplane` principal as a member end up inheriting permissions from the `flyte_data_role`? Or if this sas group not end up inheriting permissions from the data role, why do we include the data role as part of the sas group?

Reply:
`flyte_dataplane` is a member of all three groups:
- `flyte_metadata_group`
- `flyte_data_group`
- `flyte_sas_group`

and thus inherits all three roles:
- `flyte_metadata_role`
- `flyte_data_role`
- `flyte_sas_role`

`flyte_controlplane` is a member of `flyte_metadata_group` only and inherits the `flyte_metadata_role`.

There is a one-to-one binding (role assignment) between a group and a role:
- `flyte_metadata_group` <-> `flyte_metadata_role`
- `flyte_data_group` <-> `flyte_data_role`
- `flyte_sas_group` <-> `flyte_sas_role`

The `flyte_data_role` and `flyte_sas_role` are separate because they have different scopes:
- `flyte_data_role` - scoped to the data storage container
- `flyte_sas_role` - scoped to the storage account

Here's a screenshot for the flyte_dataplane role assignment configuration in the Azure portal:

<img width="1153" alt="Screenshot 2024-03-26 at 1 16 31 PM" src="https://github.com/dominodatalab/terraform-azure-aks/assets/43554115/786d83cf-2b57-4c8f-a4d8-42577b7795a2">

Reply:
I see, thanks for the detailed explanation

Reply:
np, it could be a bit confusing reading the Terraform code :)